libtenzir/include/tenzir/secret.hpp

```diff
@@ -0,0 +1,213 @@
+  }
+  if (std::equal(l.begin(), l.end(), r.begin())) {
+    return std::partial_ordering::equivalent;
+  }
+  return std::partial_ordering::unordered;
+}
+
+template <typename T1, typename T2>
+auto operator==(const secret_common<T1>& lhs, const secret_common<T2>& rhs)
+  -> bool {
+  return (lhs <=> rhs) == std::partial_ordering::equivalent;
+}
+
+/// If we dont manually implement this, we run into some issues with the
+/// recursive `hash_inspector`.
+template <class HashAlgorithm, typename FlatbufferType>
+void hash_append(HashAlgorithm& h, const secret_common<FlatbufferType>& s) {
+  return hash_append(h, s.bytes());
+}
+
+} // namespace detail::secrets
+
+/// @relates detail::secret::secret_common
+class secret final
+  : public detail::secrets::secret_common<detail::secrets::owning_fbs_buffer> {
+public:
+  using impl
+    = detail::secrets::secret_common<detail::secrets::owning_fbs_buffer>;
+  friend class secret_view;
+  using impl::impl;
+
+  secret(const impl& base) : impl{base} {
+  }
+  secret(impl&& base) : impl{std::move(base)} {
+  }
+
+  secret(std::string_view name, std::string_view operations, bool is_literal);
+
+  static auto make_literal(std::string_view value) -> secret;
+  static auto make_managed(std::string_view value) -> secret;
+  static auto from_fb(const fbs::data::Secret*) -> secret;
+};
+
+/// @relates detail::secret::secret_common
+/// TODO: Currently a `secret_view` is identical to a `secret`, because both use
+/// the owning tenzir::flatbuffer wrapper. We ideally want a non-owning version
+/// of `flatbuffer` that does not hold a `chunk_ptr`, but only the `Table*`.
+class secret_view final
+  : public detail::secrets::secret_common<detail::secrets::viewing_fbs_buffer> {
+public:
+  using impl
+    = detail::secrets::secret_common<detail::secrets::viewing_fbs_buffer>;
+  using impl::impl;
+
+  secret_view(const impl& base) : impl{base} {
+  }
+  secret_view(impl&& base) : impl{std::move(base)} {
+  }
+  secret_view(const secret& s);
+};
+
+inline auto materialize(secret_view v) -> secret {
+  return secret{v.buffer};
+}
+
+template <class HashAlgorithm>
+void hash_append(HashAlgorithm& h, const secret& s) {
+  return hash_append(h, static_cast<const secret::impl&>(s));
+}
+
+template <class HashAlgorithm>
+void hash_append(HashAlgorithm& h, const secret_view& s) {
+  return hash_append(h, static_cast<const secret_view::impl&>(s));
+}
+
+template <class T>
+struct secret_printer : printer_base<secret_printer<T>> {
+  using attribute = T;
+
+  template <class Iterator>
+  bool print(Iterator& out, const attribute& x) const {
+    return x.print_to(out);
+  }
+};
+
+template <>
+struct printer_registry<secret> {
+  using type = secret_printer<secret>;
+};
+
+template <>
+struct printer_registry<secret_view> {
+  using type = secret_printer<secret_view>;
+};
+
+} // namespace tenzir
+
+namespace fmt {
+
+template <typename BlobType>
+struct formatter<tenzir::detail::secrets::secret_common<BlobType>> {
```

Comment:
Since this allows for formatting the internals of a secret, it makes it easy to accidentally use and pass around this internal representation and then format it. Ideally, that should be forbidden, and `secret` and `secret_view` should be formattable directly.